stan/math/prim/err/check_flag_sundials.hpp

```diff
@@ -15,34 +181,61 @@ namespace math {
  * @param func_name Name of the function that returned the flag
  * @throw <code>std::runtime_error</code> if the flag is negative
  */
-inline void check_flag_sundials(int flag, const char* func_name) {
+inline void cvodes_check(int flag, const char* func_name) {
   if (flag < 0) {
     std::ostringstream ss;
-    ss << func_name << " failed with error flag " << flag << ".";
+    ss << func_name << " failed with error flag " << flag << ": \n"
+       << cvodes_flag_msg(flag).at(1) << ".";
+    if (flag == -1 || flag == -4) {
+      throw std::domain_error(ss.str());
+    } else {
+      throw std::runtime_error(ss.str());
+    }
+  }
+}
+
+/**
+ * Throws an exception message when the functions in KINSOL
+ * fails. When the exception is caused
+ * by a tuning parameter the user controls, gives a specific
+ * error. "KINGetReturnFlagName()" from sundials has a mem leak bug so
+ * until it's fixed we cannot use it to extract flag error string.
+ *
+ * @param flag Error flag
+ * @param func_name calling function name
+ * @throw <code>std::runtime_error</code> if the flag is negative.
+ */
+inline void kinsol_check(int flag, const char* func_name) {
+  std::ostringstream ss;
+  if (flag < 0) {
+    ss << "algebra_solver failed with error flag " << flag << ".";
     throw std::runtime_error(ss.str());
   }
 }
 
 /**
- * Throws an exception message when the function KINSol()
- * (call to the solver) fails. When the exception is caused
+ * Throws an exception message when the KINSol() call fails.
+ * When the exception is caused
  * by a tuning parameter the user controls, gives a specific
  * error.
  *
  * @param flag Error flag
- * @param max_num_steps Maximum number of iterations the algebra solver
- *   should take before throwing an error
- * @throw <code>std::domain_error</code> if flag means maximum number of
- *   iterations exceeded in the algebra solver.
- * @throw <code>std::runtime_error</code> if the flag is negative for
- *   any other reason.
+ * @param func_name calling function name
+ * @param max_num_steps max number of nonlinear iters
```

Comment:
Let's just use one of the terms steps/iterations.

Reply:
changed to "iterations".